src/algebra/linear_recurrence.lean

```diff
@@ -0,0 +1,213 @@
+      by rw eq_mk_of_is_sol_of_eq_init
+  end
+
+/-- If `u` is a solution to `E` and `init` designates its first `E.order` values,
+  then `u = E.mk_sol init`. This proves that `E.mk_sol init` is the only solution
+  of `E` whose first `E.order` values are given by `init`. -/
+lemma eq_mk_of_is_sol_of_eq_init' {u : ℕ → α} {init : fin E.order → α}
+  (h : E.is_solution u) (heq : ∀ n : fin E.order, u n = init n) :
+  u = E.mk_sol init := funext (E.eq_mk_of_is_sol_of_eq_init h heq)
+
+example : semimodule α (ℕ → α) := by apply_instance
+
+/-- The space of solutions of `E`, as a `submodule` over `α` of the semimodule `ℕ → α`. -/
+def sol_space : submodule α (ℕ → α) :=
+{ carrier := {u | E.is_solution u},
+  zero_mem' := λ n, by simp,
+  add_mem' := λ u v hu hv n, by simp [mul_add, sum_add_distrib, hu n, hv n],
+  smul_mem' := λ a u hu n, by simp [hu n, mul_sum]; congr'; ext; ac_refl }
+
+/-- Defining property of the solution space : `u` is a solution
+  iff it belongs to the solution space. -/
+lemma is_sol_iff_mem_sol_space (u : ℕ → α) :
+  E.is_solution u ↔ u ∈ E.sol_space := by refl
+
+/-- The function that maps a solution `u` of `E` to its first
+  `E.order` terms as a `linear_equiv`. -/
+def to_init :
+  E.sol_space ≃ₗ[α] (fin E.order → α) :=
+{ to_fun := λ u x, (u : ℕ → α) x,
+  map_add' := λ u v, by ext; simp; norm_cast,
+  map_smul' := λ a u, by ext; simp,
+  inv_fun := λ u, ⟨E.mk_sol u, E.is_sol_mk_sol u⟩,
+  left_inv := λ u, by ext n; symmetry; apply E.eq_mk_of_is_sol_of_eq_init u.2; intros k; refl,
+  right_inv := λ u, function.funext_iff.mpr (λ n, E.mk_sol_eq_init u n) }
+
+/-- Two solutions are equal iff they are equal on `range E.order`. -/
+lemma sol_eq_of_eq_init (u v : ℕ → α) (hu : E.is_solution u) (hv : E.is_solution v) :
+  u = v ↔ set.eq_on u v ↑(range E.order) :=
+begin
+  refine iff.intro (λ h x hx, h ▸ rfl) _,
+  intro h,
+  set u' : ↥(E.sol_space) := ⟨u, hu⟩,
+  set v' : ↥(E.sol_space) := ⟨v, hv⟩,
+  change u'.val = v'.val,
+  suffices h' : u' = v', from h' ▸ rfl,
+  rw [← E.to_init.to_equiv.apply_eq_iff_eq, linear_equiv.coe_to_equiv],
+  unfold_coes,
+  simp only [linear_recurrence.to_init],
+  ext x,
+  exact_mod_cast h (mem_range.mpr x.2)
+end
+
+/-! `E.tuple_succ` maps `![s₀, s₁, ..., sₙ]` to `![s₁, ..., sₙ, ∑ (E.coeffs i) * sᵢ]`,
+  where `n := E.order`. This operation is quite useful for determining closed-form
+  solutions of `E`. -/
+
+/-- `E.tuple_succ` maps `![s₀, s₁, ..., sₙ]` to `![s₁, ..., sₙ, ∑ (E.coeffs i) * sᵢ]`,
+  where `n := E.order`. -/
+def tuple_succ : (fin E.order → α) →ₗ[α] (fin E.order → α) :=
+{ to_fun := λ X i, if h : i.1 + 1 < E.order then X ⟨i+1, h⟩ else (∑ i, E.coeffs i * X i),
+  map_add' := λ x y,
+    begin
+      ext i,
+      by_cases h : i.1 + 1 < E.order,
+      { simp [h] },
+      { simp [h, mul_add, sum_add_distrib] }
+    end,
+  map_smul' := λ x y,
+    begin
+      ext i,
+      by_cases h : i.1 + 1 < E.order,
+      { simp [h] },
+      { simp [h, mul_sum], congr, ext, ac_refl }
```

Comment:
```suggestion
    begin
      ext i,
      split_ifs ; simp [h, mul_add, sum_add_distrib],
    end,
  map_smul' := λ x y,
    begin
      ext i,
      split_ifs ; simp [h, mul_sum],
      exact sum_congr rfl (λ x _, by ac_refl),
    end
```
I wrote this mostly to illustrate the intended use of the semi-colon combinator (and also `split_ifs`).